### ISSUE TITLE
Repair segment retains RUNNING state following segment repair failure and Cassandra backend timeout

### ISSUE THREAD

**Issue by sergeyyegournov** [thelastpickle/cassandra-reaper#320]
Running the latest 1.1.0-SNAPSHOT with Cassandra backend. I have a repair which is stuck (2 segments remaining).

The reason for that is that these two segments are in the RUNNING state according to the repair_run table:
```

cqlsh> select segment_id from reaper_db.repair_run where id = 40aee550-0069-11e8-9208-931ae1c16c5b and segment_state = 1 allow filtering

 segment_id
--------------------------------------
 40aee59c-0069-11e8-9208-931ae1c16c5b
 40b0451b-0069-11e8-9208-931ae1c16c5b

```
 
However there are no segment runners for them. Hence when `RepairRunner.startNextSegment` runs it logs `Running segment for range` and then `context.storage.getNextFreeSegmentInRange(repairRunId, Optional.of(parallelRanges.get(rangeIndex)))` does not return any segment because it returns segments in the NOT_STARTED state only.

According to the logs this is what happened to one of these segments stuck in RUNNING state:

1.  Segment repair (40aee59c-0069-11e8-9208-931ae1c16c5b) gets started:
`INFO   [2018-01-25 11:30:20,840] [paidrecs:40aee550-0069-11e8-9208-931ae1c16c5b] i.c.s.RepairRunner - Next segment to run : 40aee59c-0069-11e8-9208-931ae1c16c5b
INFO   [2018-01-25 11:30:52,705] [paidrecs:40aee550-0069-11e8-9208-931ae1c16c5b:40aee59c-0069-11e8-9208-931ae1c16c5b] i.c.s.SegmentRunner - It is ok to repair segment '40aee59c-0069-11e8-9208-931ae1c16c5b' on repair run with id '40aee550-0069-11e8-9208-931ae1c16c5b'
`
2. Segment repair fails and backend update times out:
`
ERROR  [2018-01-25 11:31:24,082] [paidrecs:40aee550-0069-11e8-9208-931ae1c16c5b:40aee59c-0069-11e8-9208-931ae1c16c5b] i.c.s.RepairRunner - Executing SegmentRunner failed
com.datastax.driver.core.exceptions.OperationTimedOutException: [/10.40.66.47:9042] Timed out waiting for server response
        at com.datastax.driver.core.exceptions.OperationTimedOutException.copy(OperationTimedOutException.java:44)
        at com.datastax.driver.core.exceptions.OperationTimedOutException.copy(OperationTimedOutException.java:26)
        at com.datastax.driver.core.DriverThrowables.propagateCause(DriverThrowables.java:37)
        at com.datastax.driver.core.DefaultResultSetFuture.getUninterruptibly(DefaultResultSetFuture.java:245)
        at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:68)
        at io.cassandrareaper.storage.CassandraStorage.hasLeadOnSegment(CassandraStorage.java:1046)
        at io.cassandrareaper.storage.CassandraStorage.updateRepairSegment(CassandraStorage.java:618)
        at io.cassandrareaper.service.SegmentRunner.processTriggeredSegment(SegmentRunner.java:299)
        at io.cassandrareaper.service.SegmentRunner.runRepair(SegmentRunner.java:277)
        at io.cassandrareaper.service.SegmentRunner.run(SegmentRunner.java:133)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:108)
        at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:41)
        at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:77)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: com.datastax.driver.core.exceptions.OperationTimedOutException: [/10.40.66.47:9042] Timed out waiting for server response
        at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onTimeout(RequestHandler.java:772)
        at com.datastax.driver.core.Connection$ResponseHandler$1.run(Connection.java:1374)
        at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:581)
        at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:656)
        at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:367)
`

Since then there is nothing about 40aee59c-0069-11e8-9208-931ae1c16c5b segment in the logs - and the segment is left in the RUNNING state in the storage without a segment runner instance for it.

**Comment by adejanovski** [comment]
Hi @sergeyyegournov,

excellent analysis, thanks ! 

I've been reproducing the issue (well kind of) and here's what seem to be happening : 

In the `ReaperApplication` class, a scheduler invokes the poorly named `RepairManager.resumeRunningRepairRuns()` every 10 seconds.
That method is responsible for : 

* starting `RepairRunner` instances for repairs with status == `RUNNING`
* Aborting `RUNNING` segments when the repair run has status == `PAUSED`
* Aborting `RUNNING` segments for `RUNNING` repair runs when they have no leader (ie : no Reaper instance is apparently handling them and they do not seem to be running anymore)

What seems to be happening in your case is that an uncatched exception is raised somewhere down the call chain and the scheduler thread crashes silently, since we're only catching exceptions of type `ReaperException` : https://github.com/thelastpickle/cassandra-reaper/blob/master/src/server/src/main/java/io/cassandrareaper/ReaperApplication.java#L211-L216

I've just changed the `RepairManager.resumeRunningRepairRuns()` methods to throw a ReaperException for each RuntimeException that we can catch there.
What I expect is that anytime an issue like yours happens, it will not silently crash the scheduler so that we can see in the logs what failed.
I also did so that Reaper does not send a cancellation of all running repairs on the coordinator node, since we suspect that repair might not be running anymore (and we could cancel repair of other segments by doing this).

I've created PR #321 with the fix and will release a new snapshot as soon as it is merged into master. (you can build it in the meantime to test it).

Thanks again !

**Comment by sergeyyegournov** [comment]
@adejanovski, thank you for kind words. I totally missed existence of `RepairManager.resumeRunningRepairRuns()` - and actually when I thought on how to fix the issue I have faced is that it would be nice to have another runner scheduled which will handle issues like segments left in the RUNNING state in my case - but apparently it's in there already :-) Thank you for quick turnaround - Reaper is just amazing and the only real solution to running repairs on large production clusters so it's a pleasure to make discover and report issues in the effort to make it even better!

**Comment by sergeyyegournov** [comment]
@adejanovski I built reaper from the branch with PR and I can confirm that once I started reaper these segments which were left in the RUNNING state got repaired successfully. By the way this time I did not find any exception trace in the log for `RepairManager.resumeRunningRepairRuns()`so it succeeded from first try anyway (with the older build restarting reaper did not help so clearly something went wrong with `RepairManager.resumeRunningRepairRuns()`)

**Comment by gaborauth** [comment]
Sometimes (one out of ~50)  one or more repair stuck with the following repeated log message (with 1.0.0 version), is it relevant with this (fixed) issue? A restart of cassandra-reaper resolve it.

`INFO   [2018-02-20 08:24:07,426] [keyspace:91da86a0-15bd-11e8-992f-53e0d3148ab7] i.c.s.RepairRunner - Running segment for range (-9156128396985454027,-9156128396985454027]`
`INFO   [2018-02-20 08:24:07,523] [keyspace:91da86a0-15bd-11e8-992f-53e0d3148ab7] i.c.s.RepairRunner - Repair amount done 375`
`INFO   [2018-02-20 08:24:18,097] [keyspace:91da86a0-15bd-11e8-992f-53e0d3148ab7] i.c.s.RepairRunner - Running segment for range (-9156128396985454027,-9156128396985454027]`
`INFO   [2018-02-20 08:24:18,197] [keyspace:91da86a0-15bd-11e8-992f-53e0d3148ab7] i.c.s.RepairRunner - Repair amount done 375`
`INFO   [2018-02-20 08:24:28,688] [keyspace:91da86a0-15bd-11e8-992f-53e0d3148ab7] i.c.s.RepairRunner - Running segment for range (-9156128396985454027,-9156128396985454027]`
`INFO   [2018-02-20 08:24:28,778] [keyspace:91da86a0-15bd-11e8-992f-53e0d3148ab7] i.c.s.RepairRunner - Repair amount done 375`
`INFO   [2018-02-20 08:24:39,263] [keyspace:91da86a0-15bd-11e8-992f-53e0d3148ab7] i.c.s.RepairRunner - Running segment for range (-9156128396985454027,-9156128396985454027]`
`INFO   [2018-02-20 08:24:39,357] [keyspace:91da86a0-15bd-11e8-992f-53e0d3148ab7] i.c.s.RepairRunner - Repair amount done 375`

**Comment by adejanovski** [comment]
@gaborauth, yes it is related.
We still have one case where a segment runner fails to be cleared upon exception and that can hang repairs, which I plan to fix today or tomorrow.
I suggest to upgrade to the latest beta that contain the fix, or wait for the 1.1.0 release that should ship this week or early next week.

**Comment by gaborauth** [comment]
@adejanovski, thank you! I can wait for the 1.1.0 release, it is not blocker issue for me... :)
